### PR TITLE
[MKM] Implement Anzrag's Rampage; add hint to Structural Assault

### PR DESCRIPTION
**Pull request by Cguy7777**
Since Anzrag's Rampage's `Watcher` is identical to Structural Assault's, it seemed fitting to add the same hint over there.

### CODE REVIEW

**Review by xenohedron — APPROVED** [review]

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    private static final ValueHint hint = new ValueHint(
+            "Artifacts put into graveyards from the battlefield this turn", AnzragsRampageValue.instance);
+
+    public AnzragsRampage(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.SORCERY}, "{3}{R}{R}");
+
+        // Destroy all artifacts you don't control, then exile the top X cards of your library,
+        // where X is the number of artifacts that were put into graveyards from the battlefield this turn.
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/a/AnzragsRampage.java:50]: Do not split single action by multiple lines in card rule comments. It’s important to keep it in same line for better code searching/regexp.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        super(card);
+    }
+
+    @Override
+    public AnzragsRampage copy() {
+        return new AnzragsRampage(this);
+    }
+}
+
+// CardsPutIntoGraveyardWatcher does not count tokens so custom watcher is needed.
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/a/AnzragsRampage.java:69]: No need ref class comments for simple code like watcher.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        TargetCard targetCard = new TargetCardInExile(
+                0, 1, StaticFilters.FILTER_CARD_CREATURE, CardUtil.getExileZoneId(game, source));
+        targetCard.withNotTarget(true);
+        controller.choose(outcome, targetCard, source, game);
+        Card card = game.getCard(targetCard.getFirstTarget());
+        if (card == null) {
+            return true;
+        }
+
+        if (controller.moveCards(card, Zone.BATTLEFIELD, source, game)) {
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/a/AnzragsRampage.java:160]: If you use code with multiple zones moving then make sure it works with MDFC cards and on replacement effects like commander return.

Possible bugs/checking while testing:
- second delayed part works after blink, possible fix: miss mor usage somewhere;
- second part works after zone replacement applied (commander move, another effect), possible fix: card zone check before second part call;
- second part does not works (same replacement effect - see above), possible fix: miss state.ProcessActions after move to exile/grave/hand zones (move to battlefield works fine);

Comment (JayDi85) [reply Mage.Sets/src/mage/cards/a/AnzragsRampage.java:160]: ^ code must be fine here cause it uses move to battlefield and new permanent after it. Just keep in mind about possible problems.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    }
+}
+
+class AnzragsRampageEffect extends OneShotEffect {
+
+    AnzragsRampageEffect() {
+        super(Outcome.PutCreatureInPlay);
+        staticText = "exile the top X cards of your library, where X is the number of artifacts " +
+                "that were put into graveyards from the battlefield this turn. " +
+                "You may put a creature card exiled this way onto the battlefield. It gains haste. " +
```
Comment (JayDi85) [inline Mage.Sets/src/mage/cards/a/AnzragsRampage.java:129]: Its’s a big effect with multiple steps/actions, so add related card rules in the code blocks as comments for easier review and modify.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java (hunk 3 above)
Comment (Cguy7777) [reply Mage.Sets/src/mage/cards/a/AnzragsRampage.java:160]: Okay, awesome. I had started testing scenarios you mentioned, and everything was good so far:

- If I put a creature onto the battlefield and flicker it, the return-to-hand effect doesn't bounce it.
- I cast my commander and put it on top of my library with Time Ebb. Then I cast Anzrag's Rampage, and chose to put them onto the battlefield. At my end step, the return-to-hand effect triggers and I can decide if I want them in hand or in the command zone.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java (hunk 1 above)
Comment (Cguy7777) [reply Mage.Sets/src/mage/cards/a/AnzragsRampage.java]: I think I've rearranged it to be better, but please tell me if the new singular line break would still be an issue.

File: Mage.Sets/src/mage/cards/a/AnzragsRampage.java (hunk 2 above)
Comment (xenohedron) [reply Mage.Sets/src/mage/cards/a/AnzragsRampage.java]: Personally I think that is a useful comment. It's not just referencing, but explaining the reasoning behind them being separate (can protect against ill-advised refactors that miss the distinction)

**Pull request merged**